Allow subgrids to participate in masonry layout.

sammygill opened the pull request:
#### 15f5b0f32da7334c055e80173df8078ba2598262
<pre>
Allow subgrids to participate in masonry layout.
<a href="https://bugs.webkit.org/show_bug.cgi?id=248403">https://bugs.webkit.org/show_bug.cgi?id=248403</a>
rdar://102720861

Reviewed by Brent Fulgham.

Subgrids should be able to participate in masonry layout in the same
axis as their parent grids. This patch allows this to happen by&apos;
modifying isMasonryRows/Columns to return true when the parent grid is
participating in masonry in the appropriate axis and if the same axis is
being subgridded. In these scenarios, isSubgrid(ForRows/ForColumns) will
need to return false because the axis cannot be subgridded in the same
sense. However, the grid axis can still be subgridded normally.

Also, if a subgrid is being auto-placed, then it cannot inherit any of
its parent grid&apos;s line names. NamedLineCollection has been modified to
check if a subgrid is participating in masonry layout via isMasonryRows
and isMasonryColumns. It then checks to see if the grid axis is being
automatically positioned. If both of these are true, then it will exit
early and not inherit any line names.

Spec reference: <a href="https://drafts.csswg.org/css-grid-3/#subgrids">https://drafts.csswg.org/css-grid-3/#subgrids</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-columns-subgrid-named-lines-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-columns-subgrid-named-lines-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-columns-subgrid-named-lines-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-columns-subgrid-named-lines-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-columns-subgrid-named-lines-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-columns-subgrid-named-lines-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-rows-subgrid-named-lines-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-rows-subgrid-named-lines-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-rows-subgrid-named-lines-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-rows-subgrid-named-lines-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-rows-subgrid-named-lines-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-rows-subgrid-named-lines-003.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/rendering/GridMasonryLayout.cpp:
(WebCore::GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition):
(WebCore::GridMasonryLayout::setItemGridAxisContainingBlockToGridArea):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::computeTrackBasedSize const):
(WebCore::GridTrackSizingAlgorithm::isDirectionInMasonryDirection const):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::sizesLogicalWidthToFitContent const):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutBlock):
(WebCore::RenderGrid::isMasonryRows const):
(WebCore::RenderGrid::isMasonryColumns const):
(WebCore::RenderGrid::isSubgrid const):
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/style/GridPositionsResolver.cpp:
(WebCore::NamedLineCollection::NamedLineCollection):

Canonical link: <a href="https://commits.webkit.org/257916@main">https://commits.webkit.org/257916@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5955738683499555a39d454b7cbe31ecce7833e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100355 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9517 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33418 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109672 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169891 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104348 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10429 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/125 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92756 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107550 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106133 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7894 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91159 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34542 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22551 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3258 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24068 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3251 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9372 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43558 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5437 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5067 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/rendering/GridMasonryLayout.cpp:
```diff
@@ -162,6 +163,7 @@ void GridMasonryLayout::setItemGridAxisContainingBlockToGridArea(RenderBox& chil
         child.setOverridingContainingBlockContentLogicalWidth(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(child, ForColumns));
     else
         child.setOverridingContainingBlockContentLogicalHeight(m_renderGrid.m_trackSizingAlgorithm.estimatedGridAreaBreadthForChild(child, ForRows));
+    child.setChildNeedsLayout(MarkOnlyThis);
```
Nit: [Future Optimization Idea] Could we avoid dirtying the layout if we end up with the same containing block dimensions as before?

sammygill replied:
Yup sounds like a great idea! Created a new bug: https://bugs.webkit.org/show_bug.cgi?id=249230

brentfulgham commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -211,7 +211,8 @@ const GridTrackSize& GridTrackSizingAlgorithm::rawGridTrackSize(GridTrackSizingD
 {
-    if (m_renderGrid->style().gridMasonryRows() || m_renderGrid->style().gridMasonryColumns())
+    // Size in the masonry direction is the masonry content size
+    if ((m_renderGrid->isMasonryRows() && m_direction == ForRows) || (m_renderGrid->isMasonryColumns() && m_direction == ForColumns))
```
You could remove this comment if you created a helper method:

```
bool isTrackInMasonryDirection() const
{
    return (m_renderGrid->isMasonryRows() && m_direction == ForRows) || (m_renderGrid->isMasonryColumns() && m_direction == ForColumns);
}
```

sammygill replied:
Done

brentfulgham commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -2812,7 +2812,7 @@ bool RenderBox::sizesLogicalWidthToFitContent(SizeType widthType) const
 
     if (isGridItem())
-        return !hasStretchedLogicalWidth();
+        return downcast<RenderGrid>(parent())->isMasonryColumns() || !hasStretchedLogicalWidth();
```
Is it possible for this downcast to fail? It looks like parent() is nullptr-checked in other RenderBox methods.

Maybe safer to do:

```
if (isGridItem())
    return parent() && downcast<RenderGrid>(parent())->isMasonryColumns() || !hasStretchedLogicalWidth();
```

sammygill replied:
`isGridItem()` seems to check `parent()` so I don't think it should fail. Should I still add the check or do you think it's fine?

brentfulgham replied:
If `isGridItem()` checks `parent()`, no need to add my suggested change. Thanks!

brentfulgham commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -698,17 +701,24 @@ static GridArea insertIntoGrid(Grid& grid, RenderBox& child, const GridArea& are
 }
 
-// TODO: Update name of functions to 'areMasonryRows'.
+// FIXME: Change isMasonryRows/Columns to areMasonryRows/Columns (249181)
```
Nit: FIXME(249181): ...

sammygill replied:
Fixed

stwrt replied:
I threw up a quick pr to fix this https://github.com/WebKit/WebKit/pull/7625.

This really does not fit this patch description.

stwrt replied:
You might need to rebase this patch as the name change landed.

brentfulgham reviewed (approved):
Looks like a reasonable change.

sammygill committed:
Allow subgrids to participate in masonry layout.
https://bugs.webkit.org/show_bug.cgi?id=248403
rdar://102720861

Reviewed by Brent Fulgham.

Subgrids should be able to participate in masonry layout in the same
axis as their parent grids. This patch allows this to happen by'
modifying isMasonryRows/Columns to return true when the parent grid is
participating in masonry in the appropriate axis and if the same axis is
being subgridded. In these scenarios, isSubgrid(ForRows/ForColumns) will
need to return false because the axis cannot be subgridded in the same
sense. However, the grid axis can still be subgridded normally.

Also, if a subgrid is being auto-placed, then it cannot inherit any of
its parent grid's line names. NamedLineCollection has been modified to
check if a subgrid is participating in masonry layout via isMasonryRows
and isMasonryColumns. It then checks to see if the grid axis is being
automatically positioned. If both of these are true, then it will exit
early and not inherit any line names.

Spec reference: https://drafts.csswg.org/css-grid-3/#subgrids

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-columns-subgrid-named-lines-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-columns-subgrid-named-lines-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-columns-subgrid-named-lines-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-columns-subgrid-named-lines-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-columns-subgrid-named-lines-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-columns-subgrid-named-lines-003.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-rows-subgrid-named-lines-001-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-rows-subgrid-named-lines-001.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-rows-subgrid-named-lines-002-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-rows-subgrid-named-lines-002.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-rows-subgrid-named-lines-003-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/subgrid/masonry-rows-subgrid-named-lines-003.html: Added.
* LayoutTests/platform/gtk/TestExpectations:
* Source/WebCore/rendering/GridMasonryLayout.cpp:
(WebCore::GridMasonryLayout::placeItemsWithDefiniteGridAxisPosition):
(WebCore::GridMasonryLayout::setItemGridAxisContainingBlockToGridArea):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::computeTrackBasedSize const):
(WebCore::GridTrackSizingAlgorithm::isDirectionInMasonryDirection const):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::sizesLogicalWidthToFitContent const):
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutBlock):
(WebCore::RenderGrid::isMasonryRows const):
(WebCore::RenderGrid::isMasonryColumns const):
(WebCore::RenderGrid::isSubgrid const):
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/style/GridPositionsResolver.cpp:
(WebCore::NamedLineCollection::NamedLineCollection):

Canonical link: https://commits.webkit.org/257916@main
